SmartDeviceLink/private/SDLChoiceSetManager.m

```diff
@@ -278,15 +280,48 @@ - (void)preloadChoices:(NSArray<SDLChoiceCell *> *)choices withCompletionHandler
             return;
         }
 
+        // Update the list of `preloadedMutableChoices` and `pendingMutablePreloadChoices` with the successful choice uploads
         [SDLGlobals runSyncOnSerialSubQueue:self.readWriteQueue block:^{
             __strong typeof(weakSelf) strongSelf = weakSelf;
-            [strongSelf.preloadedMutableChoices unionSet:choicesToUpload.set];
-            [strongSelf.pendingMutablePreloadChoices minusSet:choicesToUpload.set];
+            if (failedChoiceUploadSet.count == 0) {
+                [strongSelf.preloadedMutableChoices unionSet:choicesToUpload.set];
+                [strongSelf.pendingMutablePreloadChoices minusSet:choicesToUpload.set];
+            } else {
+                NSMutableSet *successfulChoiceUploads = [NSMutableSet setWithSet:choicesToUpload.set];
+                [successfulChoiceUploads minusSet:failedChoiceUploadSet];
+
+                [strongSelf.preloadedMutableChoices unionSet:[successfulChoiceUploads copy]];
+                [strongSelf.pendingMutablePreloadChoices minusSet:choicesToUpload.set];
+
+                // Add the failed choices to all the pending present operations
+                [strongSelf sdl_updatePendingOperationsWithFailedPreloadedChoices:[failedChoiceUploadSet copy]];
+            }
         }];
+    }];
+
+    __weak typeof(preloadOp) weakPreloadOp = preloadOp;
+    preloadOp.completionBlock = ^{
+        SDLLogD(@"Choices finished preloading");
+
+        if (handler != nil) {
+            handler(weakPreloadOp.error);
+        }
     };
+
     [self.transactionQueue addOperation:preloadOp];
 }
 
+/// Update currently pending preload choice operations with the failed choice commands so a re-upload can be attempted.
+/// @param failedPreloadedChoices The failed choice commands
+- (void)sdl_updatePendingOperationsWithFailedPreloadedChoices:(NSSet<SDLChoiceCell *> *)failedPreloadedChoices {
+    for (NSOperation *operation in self.transactionQueue.operations) {
+        if (operation.isExecuting || ![operation isKindOfClass:SDLPreloadChoicesOperation.class]) { continue; }
+
+        SDLPreloadChoicesOperation *updateOp = (SDLPreloadChoicesOperation *)operation;
+        [updateOp addFailedChoicesToUpload:failedPreloadedChoices];
+    }
+}
```

Comment:
I don't think this is the right approach. It's complicated, and it adds potentially unnecessary choices to all future operations. Instead:

1. If there was an attached `present`, it will fail because some of the needed choices are not available.
2. If the preload was standalone, some of the choices will have simply failed to upload. If a future present occurs attempting to use some of those choices and some others that failed, they will be re-attempted automatically then with the current setup.

I think that most of these changes, then, should be reverted.

Reply:
Fixed. Removed.